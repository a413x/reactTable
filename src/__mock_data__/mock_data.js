export const mock_data = [
  {
    "id":"1",
    "firstName":"Frederick",
    "lastName":"Catania",
    "email":"GHellems@magna.net",
    "phone":"(458)969-6847",
    "address":{"streetAddress":"7471 Porttitor Ln","city":"Gig Harbor","state":"WY","zip":"15219"},
    "description":"sit pretium ipsum consectetur eros"
  },
  {
    "id":"2",
    "firstName":"Curtis",
    "lastName":"Bias",
    "email":"CRobichaud@nec.gov",
    "phone":"(481)075-5034",
    "address":{"streetAddress":"2730 Risus Ln","city":"West Warwick","state":"NM","zip":"60799"},
    "description":"sollicitudin malesuada tellus porta neque"
  },
  {
    "id":"3",
    "firstName":"Dora",
    "lastName":"Crenshaw",
    "email":"MSanchez@at.org",
    "phone":"(959)257-7541",
    "address":{"streetAddress":"5670 At Ave","city":"Pleasant Grove","state":"PA","zip":"21878"},
    "description":"amet lacus et fringilla orci"
  },
  {
    "id":"4",
    "firstName":"Grace",
    "lastName":"French",
    "email":"JLund@vestibulum.com",
    "phone":"(976)555-7134",
    "address":{"streetAddress":"1933 Vestibulum Rd","city":"Woonsocket","state":"FL","zip":"39433"},
    "description":"velit sed tellus suspendisse at"
  },
  {
    "id":"5",
    "firstName":"Randy",
    "lastName":"Bunch",
    "email":"KGunter@lacus.io",
    "phone":"(941)539-5519"
  }
]
