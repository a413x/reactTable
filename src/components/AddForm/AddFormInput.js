import {validate} from '../../utils/validate.js'

export const AddFormInput = ({
  colName, addingChange, validChange, valid
}) => {
  const onChange = (value) => {
    let valueValid = false
    if(colName === 'id') {
      valueValid = validate('number', value)
    }else if(colName === 'firstName' || colName === 'lastName'){
      valueValid = validate('text', value)
    }else if(colName === 'email'){
      valueValid = validate('email', value)
    }else if(colName === 'phone'){
      valueValid = validate('phone', value)
    }
    validChange(colName, valueValid)
    addingChange(colName, value)
  }

  const validClass = valid ? 'is-valid' : 'is-invalid'
  return (
    <div>
      <label>{colName}</label>
      <input
        className = {'form-control ' + validClass}
        onChange = {(e) => onChange(e.target.value)}
      />
      <div className = 'invalid-feedback'>
        Please enter correct value
      </div>
    </div>
  )
}
