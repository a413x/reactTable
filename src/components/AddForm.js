import {useState} from 'react'
import {tableColumns} from './TableHead.js'
import {AddFormInput} from './AddFormInput.js'

const objectToAdd = {}
const initialInputsValid = {}
tableColumns.forEach(colObj => initialInputsValid[colObj.name] = false)

export const AddForm = ({addCallback}) => {
  const [visible, setVisible] = useState(false)
  const [inputsValid, setInputsValid] = useState(initialInputsValid)

  const addingChange = (colName, value) => objectToAdd[colName] = value
  const validChange = (colName, valid) => {
    const newInputsValid = {...inputsValid}
    newInputsValid[colName] = valid
    setInputsValid(newInputsValid)
  }
  const onAdd = () => addCallback({...objectToAdd})
  const toggleForm = () => {
    const resetInputs = {...inputsValid}
    for(let colName in resetInputs) resetInputs[colName] = false
    setInputsValid(resetInputs)
    setVisible(!visible)
  }

  let allValid = true
  for(let colName in inputsValid){ allValid &= inputsValid[colName] }

  return (
    <div className = 'container'>
      <button className = 'btn btn-info' onClick = {toggleForm}>
        Add
      </button>
      {visible &&
        <div>
          {tableColumns.map((colObj, ind) =>
            <AddFormInput
              key = {ind}
              colName = {colObj.name}
              addingChange = {addingChange}
              validChange = {validChange}
              valid = {inputsValid[colObj.name]}
            />
          )}
          <button
            className = 'btn btn-success my-3'
            onClick = {onAdd}
            disabled = {allValid ? null : true}
          >
            +
          </button>
        </div>
      }
    </div>
  )
}
