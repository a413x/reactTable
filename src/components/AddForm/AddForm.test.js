import {render, fireEvent} from '@testing-library/react'
import {AddForm} from './AddForm.js'
import {mock_data} from '../../test/mock_data.js'

describe('AddForm component tests', () => {
  const mock_addCallback = jest.fn()
  const props = {
    addCallback: mock_addCallback
  }
  const objectToAdd = {...mock_data[4]}

  const setAllInputs = (inputs) => {
    let index = 0
    for(let colName in objectToAdd){
      fireEvent.change(inputs[index], { target: {value: objectToAdd[colName]} })
      index++
    }
  }

  test('renders properly', () => {
    const {container} = render(<AddForm {...props} />)
    expect(container).toMatchSnapshot()
  })

  test('renders properly with expanded form', () => {
    const {container, getByText} = render(<AddForm {...props} />)
    fireEvent.click(getByText('Add'))
    expect(getByText('+')).toBeDisabled()
    expect(container).toMatchSnapshot()
  })

  test('adding button is disabled if at least one input has the wrong value', () => {
    const {getByText} = render(<AddForm {...props} />)
    fireEvent.click(getByText('Add'))
    const inputs = document.querySelectorAll('input')
    fireEvent.change(inputs[0], {
      target: {value: objectToAdd.id }
    })
    fireEvent.change(inputs[1], {
      target: {value: objectToAdd.firstName }
    })
    expect(getByText('+')).toBeDisabled()
  })

  test('adding button is enabled when all inputs has correct value', () => {
    const {getByText} = render(<AddForm {...props} />)
    fireEvent.click(getByText('Add'))
    const inputs = document.querySelectorAll('input')
    setAllInputs(inputs)
    expect(getByText('+')).not.toBeDisabled()
  })

  test('adding button calls callback function with proper value', () => {
    const {getByText} = render(<AddForm {...props} />)
    fireEvent.click(getByText('Add'))
    const inputs = document.querySelectorAll('input')
    setAllInputs(inputs)
    fireEvent.click(getByText('+'))
    expect(mock_addCallback).toHaveBeenCalledWith(objectToAdd)
  })
})
