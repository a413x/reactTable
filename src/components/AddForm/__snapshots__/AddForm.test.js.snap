// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddForm component tests renders properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="add-form"
      >
        <button
          class="btn btn-info"
        >
          Add
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="add-form"
    >
      <button
        class="btn btn-info"
      >
        Add
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`AddForm component tests renders properly with expanded form 1`] = `
<div>
  <div
    class="add-form"
  >
    <button
      class="btn btn-info"
    >
      Add
    </button>
    <div>
      <div>
        <label>
          id
        </label>
        <input
          class="form-control is-invalid"
        />
        <div
          class="invalid-feedback"
        >
          Please enter correct value
        </div>
      </div>
      <div>
        <label>
          firstName
        </label>
        <input
          class="form-control is-invalid"
        />
        <div
          class="invalid-feedback"
        >
          Please enter correct value
        </div>
      </div>
      <div>
        <label>
          lastName
        </label>
        <input
          class="form-control is-invalid"
        />
        <div
          class="invalid-feedback"
        >
          Please enter correct value
        </div>
      </div>
      <div>
        <label>
          email
        </label>
        <input
          class="form-control is-invalid"
        />
        <div
          class="invalid-feedback"
        >
          Please enter correct value
        </div>
      </div>
      <div>
        <label>
          phone
        </label>
        <input
          class="form-control is-invalid"
        />
        <div
          class="invalid-feedback"
        >
          Please enter correct value
        </div>
      </div>
      <button
        class="btn btn-success my-3"
        disabled=""
      >
        +
      </button>
    </div>
  </div>
</div>
`;
